main/lsp/requests/document_symbol.cc

```diff
@@ -12,174 +13,172 @@ using namespace std;
+            for (auto member : singleton.data(ctx)->members()) {
+                if (member.second.isClassOrModule() || member.second.isMethod()) {
+                    continue;
+                }
 
-std::optional<RangeInfo> rangesForSymbol(const core::GlobalState &gs, core::SymbolRef symRef,
-                                         const UnorderedMap<core::SymbolRef, ASTSymbolInfo> &defMapping,
-                                         const UnorderedMap<core::SymbolRef, core::SymbolRef> &forced) {
-    RangeInfo info;
-    auto it = defMapping.find(symRef);
-    if (it != defMapping.end()) {
-        info.range = Range::fromLoc(gs, it->second.loc);
-        info.selectionRange = Range::fromLoc(gs, it->second.declLoc);
-    } else {
-        auto it = forced.find(symRef);
-        if (it != forced.end()) {
-            return rangesForSymbol(gs, it->second, defMapping, forced);
+                this->addChild(ctx, member.second);
+            }
         }
+    }
 
-        auto loc = symRef.loc(gs);
-        if (!loc.file().exists()) {
-            return nullopt;
+    void postTransformClassDef(core::Context ctx, ast::ExpressionPtr &expr) {
+        auto &klass = ast::cast_tree_nonnull<ast::ClassDef>(expr);
+        if (klass.symbol == core::Symbols::root()) {
+            return;
         }
 
-        info.range = Range::fromLoc(gs, loc);
-        info.selectionRange = Range::fromLoc(gs, loc);
+        fast_sort(this->scope->children,
+                  [](auto &left, auto &right) { return left->loc.beginPos() < right->loc.beginPos(); });
+
+        this->scope = this->scope->parent;
     }
 
-    if (info.range == nullptr || info.selectionRange == nullptr) {
-        return nullopt;
+    void preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &expr) {
+        auto &method = ast::cast_tree_nonnull<ast::MethodDef>(expr);
+
+        // We avoid using `this->addChild` here because we know the method occurs within `this->scope`, and we also want
+        // to give it a different `declLoc` and propagate `isAttrBestEffortUIOnly`.
+        auto node = std::make_shared<Node>(this->scope, method.symbol, ctx.locAt(method.loc));
+        node->declLoc = ctx.locAt(method.declLoc);
+        node->isAttrBestEffortUIOnly = method.flags.isAttrBestEffortUIOnly;
+        this->scope->children.emplace_back(node);
     }
 
-    return info;
-}
+    static std::vector<std::unique_ptr<DocumentSymbol>> documentSymbol(core::Context ctx, ast::ExpressionPtr &tree) {
+        core::Loc fullFile{ctx.file, core::LocOffsets{0, static_cast<uint32_t>(ctx.file.data(ctx).source().size())}};
 
-bool symbolIsAttr(core::SymbolRef symRef, const UnorderedMap<core::SymbolRef, ASTSymbolInfo> &defMapping) {
-    auto it = defMapping.find(symRef);
-    return it != defMapping.end() && it->second.isAttrBestEffortUIOnly;
-}
+        Outliner outliner;
+        outliner.scope = std::make_shared<Node>(nullptr, core::Symbols::root(), fullFile);
```

Comment:
Why not set this in the constructor?

Reply:
Could definitely do that instead of adding it here 👍